src/hooks/cgroups/pbs_cgroups.PY

```diff
@@ -1955,6 +1923,114 @@ class NodeConfig(object):
                    (caller_name(), host_resc_avail))
         return True
 
+    def take_node_offline(self):
+        """
+        Take the local node and associated vnodes offline
+        """
+        pbs.logmsg(pbs.EVENT_DEBUG4, '%s: Method called' % caller_name())
+        pbs.logmsg(pbs.EVENT_DEBUG2, '%s: Taking vnode(s) offline' %
+                   caller_name())
+        # Attempt to take vnodes that match this host offline
+        # Assume vnode names resemble self.hostname[#]
+        match_found = False
+        for vnode_name in pbs.event().vnode_list:
+            if ((vnode_name == self.hostname) or
+                    re.match(self.hostname + r'\[.*\]', vnode_name)):
+                pbs.event().vnode_list[vnode_name].state = pbs.ND_OFFLINE
+                pbs.event().vnode_list[vnode_name].comment = self.offline_msg
+                pbs.logmsg(pbs.EVENT_DEBUG2, '%s: %s; offlining %s' %
+                           (caller_name(), self.offline_msg, vnode_name))
+                match_found = True
+        if not match_found:
+            pbs.logmsg(pbs.EVENT_DEBUG2, '%s: No vnodes match %s' %
+                       (caller_name(), self.hostname))
+            return
+        # Write a file locally to reduce server traffic when the node
+        # is brought back online
+        pbs.logmsg(pbs.EVENT_DEBUG4, '%s: Offline file: %s' %
+                   (caller_name(), self.offline_file))
+        if os.path.isfile(self.offline_file):
+            pbs.logmsg(pbs.EVENT_DEBUG2,
+                       '%s: Offline file already exists, not overwriting' %
+                       caller_name())
+            return
+        try:
+            # Write a timestamp so that exechost_periodic can avoid
+            # cleaning up before this event has sent updates to server
+            with open(self.offline_file, 'w') as fd:
+                fd.write(str(time.time()))
+                fd.truncate()
```

Comment:
Why truncate here? `w` will truncate the file if it already exists.

Reply:
I had tested with 'r+' manually, which requires truncate(). You are correct, it's not necessary with 'w'.